fullstack-cert/typescript-projects/digital-pet-game/src/App.tsx

```diff
@@ -0,0 +1,301 @@
+  [PetMood.Happy]: "😺", // grinning cat
+  [PetMood.Excited]: "😻", // heart-eyes cat
+  [PetMood.Content]: "😸", // smiling cat
+  [PetMood.Sad]: "😿", // crying cat
+  [PetMood.Tired]: "😽", // kissing cat (closest to sleepy)
+  [PetMood.Sick]: "🙀", // weary cat (looks unwell)
+  [PetMood.Hungry]: "😹", // joy cat (closest to hungry/funny)
+};
+
+const STORAGE_KEY = "pet";
+
+interface usePetProps {
+  isGameStarted: boolean;
+  setGameStarted: React.Dispatch<React.SetStateAction<boolean>>;
+}
+
+function usePet({ isGameStarted, setGameStarted }: usePetProps) {
+  let [pet, setPet] = useState<Pet>({
+    name: "",
+    happiness: 100,
+    hunger: 0,
+    energy: 100,
+    species: "Cat",
+  });
+
+  useEffect(() => {
+    console.log("PLEASE BORTHER!");
+    if (!isGameStarted) return;
+    console.log("Excellent, Borther.");
+
+    const interval = setInterval(() => {
+      setPet((pet) => ({
+        ...pet,
+        happiness: Math.max(pet.happiness - 5, 0),
+        hunger: Math.min(pet.hunger + 5, 100),
+      }));
+    }, 30 * 1000);
+
+    return () => clearInterval(interval);
+  }, [isGameStarted]);
+
+  function doAction(action: PetAction) {
+    switch (action) {
+      case PetAction.EAT:
+        feedPet();
+        break;
+      case PetAction.SLEEP:
+        restPet();
+        break;
+      case PetAction.PLAY:
+        playWithPet();
+        break;
+    }
+  }
+
+  function feedPet() {
+    const myFoodFill = 10;
+    setPet((pet) => ({
+      ...pet,
+      hunger: Math.max(pet.hunger - myFoodFill, 0),
+    }));
+    pet.hunger -= myFoodFill;
+  }
+
+  function playWithPet() {
+    const myHappinessIncrease = 5;
+    const energyDecrease = 5;
+    setPet((pet) => ({
+      ...pet,
+      happiness: Math.min(pet.happiness + myHappinessIncrease, 100),
+      energy: Math.max(pet.energy - energyDecrease, 0),
+    }));
+  }
+
+  function restPet() {
+    const hungerIncrease = 5;
+    const energyIncrease = 5;
+    setPet((pet) => ({
+      ...pet,
+      hunger: Math.min(pet.hunger + hungerIncrease, 100),
+      energy: Math.min(pet.energy + energyIncrease, 100),
+    }));
+  }
+
+  function savePetData() {
+    localStorage.setItem(STORAGE_KEY, JSON.stringify(pet));
+  }
+
+  function loadPetData() {
+    pet = JSON.parse(localStorage.getItem(STORAGE_KEY) || "");
+    if (typeof pet !== "undefined") {
+      setPet(pet);
+      setGameStarted(true);
+    }
+  }
+
+  return {
+    pet,
+    doAction,
+    savePetData,
+    loadPetData,
+    setName: (name: string) => setPet({ ...pet, name }),
+  };
+}
+
+export function App() {
+  const [isGameStarted, setGameStarted] = useState(false);
+
+  let { pet, doAction, savePetData, loadPetData, setName } = usePet({
+    isGameStarted,
+    setGameStarted,
+  });
+
+  let petRef = useRef<Pet>(pet);
+  useEffect(() => {
+    petRef.current = pet;
+  }, [pet]);
+
+  const [fact, setFact] = useState<string | null>("");
+
+  //TODO: Integrate with the FCC API once it's ready.
```

Comment:
is the API ready? if so, then we can remove this todo and use the API here

Reply:
hey, currently the app is not uses any api... (dont sure if it is ready, since i dont know who is working on it...)

Reply:
I believe @a2937 currently owns the API task.

Reply:
Here be the API. https://github.com/freeCodeCamp/demo-projects/pull/682

Reply:
The API PR was approved and merged in. So this TODO comment can be removed and the API should be ready to use here.